src/main/java/org/jabref/model/openoffice/rangesort/RangeSortVisual.java

```diff
@@ -0,0 +1,176 @@
+
+        if (set.size() != inputSize) {
+            throw new IllegalStateException("visualSort: set.size() != inputSize");
+        }
+
+        // collect ordered result
+        List<RangeSortable<T>> result = new ArrayList<>(set.size());
+        for (ComparableMark<RangeSortable<T>> mark : set) {
+            result.add(mark.getContent());
+        }
+
+        if (result.size() != inputSize) {
+            throw new IllegalStateException("visualSort: result.size() != inputSize");
+        }
+
+        return result;
+    }
+
+    /**
+     *  Given a location, return its position: coordinates relative to the top left position of the
+     *  first page of the document.
+     *
+     * Note: for text layouts with two or more columns, this gives the wrong order:
+     *       top-down/left-to-right does not match reading order.
+     *
+     * Note: The "relative to the top left position of the first page" is meant "as it appears on
+     *       the screen".
+     *
+     *       In particular: when viewing pages side-by-side, the top half of the right page is
+     *       higher than the lower half of the left page. Again, top-down/left-to-right does not
+     *       match reading order.
+     *
+     * @param range  Location.
+     * @param cursor To get the position, we need az XTextViewCursor.
+     *               It will be moved to the range.
+     */
+    private static Point findPositionOfTextRange(XTextRange range, XTextViewCursor cursor) {
+        cursor.gotoRange(range, false);
+        return cursor.getPosition();
+    }
+
+    /**
+     * A reference mark name paired with its visual position.
+     *
+     * Comparison is based on (Y,X,indexInPosition): vertical compared first, horizontal second,
+     * indexInPosition third.
+     *
+     * Used for sorting reference marks by their visual positions.
+     */
+    private static class ComparableMark<T> implements Comparable<ComparableMark<T>> {
+
+        private final Point position;
+        private final int indexInPosition;
+        private final T content;
+
+        public ComparableMark(Point position, int indexInPosition, T content) {
+            this.position = position;
+            this.indexInPosition = indexInPosition;
+            this.content = content;
+        }
+
+        @Override
+        public int compareTo(ComparableMark other) {
+
+            if (position.Y != other.position.Y) {
+                return position.Y - other.position.Y;
+            }
+            if (position.X != other.position.X) {
+                return position.X - other.position.X;
+            }
+            return indexInPosition - other.indexInPosition;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+
+            if (o instanceof ComparableMark) {
+                ComparableMark other = (ComparableMark) o;
+                return ((this.position.X == other.position.X)
```

Comment:
For consistency I would use also Object.equals

Reply:
This code is possibly inconsistent with `hashCode` using `position` below, while I used `position.x` and `position.y` here.
Part of my confusion comes from that I feel "equal" should consider the `content` part,
however in `compareTo` I did not want to prescribe comparability for the `content`. 

I think I implemented `equals` and `hashCode` because `Comparable` seems to require it,
in order to make it suitable as a key in `TreeMap` and `HashMap` 

Although `compareTo` appeared to me as "natural order", it is probably not that in the sense
as `Comparable` expects it. Considering right-to-left languages suggests that it is not even
natural in a non-technical sense.

Dropping `implements Comparable<ComparableMark<T>>`, `compareTo`, `equals` and `hashCode`,
adding `private static int compareTopToBottomLeftToRight(ComparableMark a, ComparableMark b)`
to `RangeSortVisual` instead